products/editor/backend/api/codebase_sync.py

```diff
@@ -0,0 +1,110 @@
+from typing import cast
+
+from rest_framework import mixins, serializers, status, viewsets
+from rest_framework.decorators import action
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from posthog.api.routing import TeamAndOrgViewSetMixin
+from posthog.auth import PersonalAPIKeyAuthentication, SessionAuthentication
+from posthog.models import User
+from products.editor.backend.chunking.exceptions import UnsupportedLanguage
+from products.editor.backend.chunking.parser import guess_language
+from products.editor.backend.models.codebase import Codebase
+from products.editor.backend.services.codebase_sync import CodebaseSyncService
+from products.editor.backend.tasks import embed_file
+
+
+class CodebaseSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Codebase
+        fields = "__all__"
+        read_only_fields = ["id", "user", "team"]
+
+    def create(self, validated_data):
+        validated_data["user"] = self.context["request"].user
+        validated_data["team"] = self.context["get_team"]()
+        return super().create(validated_data)
+
+
+class ArtifactSerializer(serializers.Serializer):
+    id = serializers.CharField()
+    type = serializers.ChoiceField(choices=["file", "dir"])
+    parent_id = serializers.CharField(required=False, default=None)
+
+
+class CodebaseSyncSerializer(serializers.Serializer):
+    tree = serializers.ListField(child=ArtifactSerializer())
```

Comment:
Correct me if I'm wrong, does this mean we calculate the Merkle tree in the editor, and upload the entire codebase at once, per branch?

Reply:
I re-read the PR description, we're sending hashes with code attached to them, sending the entire codebase at once.